### PR TITLE
Handle parquet list data corner case

### PR DESCRIPTION
**Pull request by nvdbaranec**
Fixes an issue with a particular arrangement of page data related to lists.  Specifically, it is possible for page `N` to contain "0" rows because the values for the row it is a part of start on page `N-1` and end on page `N+1`.  This was defeating logic in the decode kernel that would erroneously cause these values to be skipped.

Similar to https://github.com/rapidsai/cudf/pull/12488 this is only reproducible with data out in the wild.  In this case, we have a file that we could in theory check in to create a test with, but it is 16 MB so it's fairly large.  Looking for feedback on whether this is too big.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by nvdbaranec** [comment]
@etseidl

File: cpp/src/io/parquet/page_data.cu [hunk 1]
```diff
@@ -115,9 +115,30 @@ inline __device__ bool is_bounds_page(page_state_s* const s, size_t min_row, siz
...
+
   return ((page_begin <= begin && page_end >= begin) || (page_begin <= end && page_end >= end));
 }
 
+/**
+ * @brief Returns whether or not a page is completely contained within the specified
+ * row bounds
+ *
+ * @param s The page to be checked
+ * @param min_row The starting row index
```
Comment (bdice) [inline cpp/src/io/parquet/page_data.cu:127]: I think "start" is a more appropriate word than "minimum."
```suggestion
 * @param start_row The starting row index
```

File: cpp/src/io/parquet/page_data.cu [hunk 2]
```diff
@@ -1838,7 +1860,19 @@ __global__ void __launch_bounds__(block_size) gpuDecodePageData(
   bool const has_repetition = s->col.max_level[level_type::REPETITION] > 0;
 
   // if we have no work to do (eg, in a skip_rows/num_rows case) in this page.
-  if (s->num_rows == 0 && !(has_repetition && is_bounds_page(s, min_row, num_rows))) { return; }
+  //
+  // corner case: in the case of lists, we can have pages that contain "0" rows if the current row
```
Comment (bdice) [inline cpp/src/io/parquet/page_data.cu:1864]: Do the quotes mean something? If not, I'd remove.
```suggestion
  // corner case: in the case of lists, we can have pages that contain 0 rows if the current row
```

File: cpp/src/io/parquet/page_data.cu [hunk 3]
```diff
@@ -1838,7 +1860,19 @@ __global__ void __launch_bounds__(block_size) gpuDecodePageData(
...
+  // starts before this page and ends after this page:
+  //       P0        P1        P2
+  //  |---------|---------|----------|
+  //        ^------------------^
+  //      row start           row end
+  // P1 will contain "0" rows
```
Comment (bdice) [inline cpp/src/io/parquet/page_data.cu:1870]:
```suggestion
  // P1 will contain 0 rows
```

Comment (nvdbaranec) [reply cpp/src/io/parquet/page_data.cu:1870]: The idea with the quotes was to imply zero-but-not-really.

Comment (bdice) [reply cpp/src/io/parquet/page_data.cu:1870]: I'm not sure if I understand the "but not really" portion. That seems important to understand the scope of the bug. Can you elaborate? I may just be missing some knowledge about the Parquet spec.

Comment (vuule) [reply cpp/src/io/parquet/page_data.cu:1870]: My understanding is that no rows start (or end) in such page. So it does not (fully) contain any rows, but it does contain data that we need to parse (and apparently we didn't, before this PR).
That's at least how I understand the ASCII art in the comment above :)

Comment (bdice) [reply cpp/src/io/parquet/page_data.cu:1870]: Got it. That helps. Thanks! I’ll approve.

Comment (bdice) [reply cpp/src/io/parquet/page_data.cu:1870]: We don’t need to remove the air-quotes, but maybe the comment can be updated with the description @vuule wrote above.

**Review by bdice — APPROVED** [review]

File: cpp/src/io/parquet/page_data.cu [hunk 4]
```diff
@@ -115,9 +115,30 @@ inline __device__ bool is_bounds_page(page_state_s* const s, size_t min_row, siz
...
+ * @brief Returns whether or not a page is completely contained within the specified
+ * row bounds
+ *
+ * @param s The page to be checked
+ * @param min_row The starting row index
+ * @param num_rows The number of rows
+ *
+ * @return True if the page is completely contained within the row bounds
+ */
+inline __device__ bool page_is_contained(page_state_s* const s, size_t min_row, size_t num_rows)
```
Comment (vuule) [inline cpp/src/io/parquet/page_data.cu:132]: I think this pattern is more common
```suggestion
inline __device__ bool is_page_contained(page_state_s* const s, size_t min_row, size_t num_rows)
```

**Review by vuule — COMMENTED** [review]
looks good, just a naming suggestion

**Review by vuule — APPROVED** [review]

**Comment by nvdbaranec** [comment]
rerun tests

**Comment by vuule** [comment]
@nvdbaranec "rerun tests" does not work any more :(
You can rerun failed jobs in the UI once the CI completes current jobs.

**Pull request merged**